File: backend/app/db/spark_ui.py
"""The Spark application UI's REST API — seeing and stopping running work.

The Thrift Server is one long-lived Spark application, so every job it runs is
visible at its application UI, whether the dashboard submitted it or somebody ran
spark-sql in the container.  This goes over HTTP rather than through the Thrift
connection on purpose: a connection busy with a query is exactly the connection
that cannot be asked about it.
"""
from datetime import datetime, timezone
from typing import Any, Dict, Iterable, List, Optional

import httpx

from app.config import settings

# Short, because the Health page polls this: a Spark UI that is slow to answer
# should leave the page saying so rather than holding the whole response up.
TIMEOUT_S = 3.0

# A job description can be a whole statement; the page wants a line, not a plan.
DESCRIPTION_LIMIT = 300


def base_url() -> str:
    return f"http://{settings.spark_ui_host}:{settings.spark_app_ui_port}"


def application_id() -> Optional[str]:
    """The Thrift Server's application, or None if the UI cannot be reached.

    The UI lists one application per JVM and the Thrift Server is the only one
    here, so the first is the right one.
    """
    response = httpx.get(f"{base_url()}/api/v1/applications", timeout=TIMEOUT_S)
    response.raise_for_status()
    applications = response.json()
    return applications[0]["id"] if applications else None


def running_jobs() -> List[Dict[str, Any]]:
    """Jobs currently running in the Thrift Server's application.

    Jobs rather than SQL executions: the jobs endpoint filters server-side, while
    the SQL one returns every execution since start-up with its whole query plan
    attached, which is megabytes to answer a question about the present.
    """
    app = application_id()
    if not app:
        return []
    response = httpx.get(
        f"{base_url()}/api/v1/applications/{app}/jobs",
        params={"status": "running"},
        timeout=TIMEOUT_S,
    )
    response.raise_for_status()
    jobs = []
    for job in response.json():
        # description is the statement the Thrift Server set for the job group;
        # name is the Spark call site, which is all there is for a job submitted
        # any other way.
        text = job.get("description") or job.get("name") or "a Spark job"
        jobs.append(
            {
                "id": str(job.get("jobId")),
                "state": (job.get("status") or "RUNNING").lower(),
                "sql": " ".join(text.split())[:DESCRIPTION_LIMIT],
                "running_s": _age_s(job.get("submissionTime")),
                "tasks_total": job.get("numTasks") or 0,
                "tasks_done": job.get("numCompletedTasks") or 0,
            }
        )
    return jobs


def _age_s(submitted: Optional[str]) -> float:
    """How long ago Spark says the job was submitted, in seconds.

    Spark stamps these as "2026-08-17T12:01:52.041GMT", which is ISO-8601 with a
    zone abbreviation where an offset belongs, so the suffix is swapped before
    parsing.  An unparseable stamp gives 0 rather than hiding the job.
    """
    if not submitted:
        return 0.0
    try:
        moment = datetime.fromisoformat(submitted.replace("GMT", "+00:00"))
    except ValueError:
        return 0.0
    return round(max(0.0, (datetime.now(timezone.utc) - moment).total_seconds()), 1)


def label(statement: str) -> str:
    """A statement as the job list reports it, for matching one against the other."""
    return " ".join(statement.split())[:DESCRIPTION_LIMIT]


def kill_jobs_for(statements: Iterable[str]) -> List[str]:
    """Kill the running jobs that are working on any of these statements.

    Matched by statement rather than killing whatever is running, because the
    Thrift Server is one application shared by everything that connects to it: a
    spark-sql session in the container would otherwise be collateral damage when a
    dashboard comparison is cancelled.

    Killing is necessary and not merely tidy.  Taking a client's connection away
    stops the dashboard waiting, but Spark carries on: HiveServer2 does not notice
    a dropped session promptly, so the job keeps its share of the cores and the
    next comparison would be timed against an orphan.
    """
    wanted = {label(statement) for statement in statements}
    killed = []
    for job in running_jobs():
        if job["sql"] in wanted:
            kill_job(job["id"])
            killed.append(job["id"])
    return killed


def kill_job(job_id: str) -> None:
    """Ask the UI to kill one job, as its own kill link does.

    Depends on spark.ui.killEnabled, which is Spark's default and is left at it.
    The handler answers with a redirect to the jobs page, so any 2xx or 3xx means
    the request was accepted; whether the job dies is then Spark's business, and
    the next poll of running_jobs() is what confirms it.
    """
    response = httpx.post(
        f"{base_url()}/jobs/job/kill/",
        params={"id": job_id},
        timeout=TIMEOUT_S,
        follow_redirects=False,
    )
    if response.status_code >= 400:
        raise RuntimeError(
            f"the Spark UI refused to kill job {job_id} (HTTP {response.status_code}); "
            "spark.ui.killEnabled must be on"
        )

File: htap-backend/src/main/java/com/thelastpickle/htap/backend/engine/SparkUi.java
package com.thelastpickle.htap.backend.engine;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.thelastpickle.htap.backend.config.SparkSettings;
import com.thelastpickle.htap.backend.support.Messages;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.Set;

/**
 * Seeing and stopping Spark work, over the application UI's REST interface.
 *
 * <p>The Thrift Server is one long-lived Spark application, so every job it runs is visible here
 * whether this backend submitted it or somebody ran {@code spark-sql} in the container. Over HTTP
 * rather than through the Thrift connection for the same reason the Presto listing is: a connection
 * busy with a query is exactly the connection that cannot be asked about it.
 */
@ApplicationScoped
public class SparkUi {

    /** The bound on one listing or one kill, matching {@link PrestoQueries#TIMEOUT} and for its reason. */
    static final Duration TIMEOUT = Duration.ofSeconds(3);

    /** A job description can be a whole statement; the page wants a line, not a plan. */
    static final int DESCRIPTION_LIMIT = 300;

    private final SparkSettings settings;
    private final Rest rest;
    private final Clock clock;

    @Inject
    SparkUi(SparkSettings settings, ObjectMapper json) {
        this(settings, json, Clock.systemUTC());
    }

    SparkUi(SparkSettings settings, ObjectMapper json, Clock clock) {
        this.settings = settings;
        this.rest = new Rest("the Spark UI", TIMEOUT, json);
        this.clock = clock;
    }

    /**
     * The Thrift Server's application, or empty when the UI has none.
     *
     * <p>The UI lists one application per JVM and the Thrift Server is the only one in that
     * container, so the first is the right one.
     */
    public Optional<String> applicationId() {
        JsonNode applications = rest.get(url("/api/v1/applications"));
        return applications.isEmpty()
                ? Optional.empty()
                : Optional.of(applications.get(0).path("id").asText(""));
    }

    /**
     * The jobs the application is running.
     *
     * <p>Jobs rather than SQL executions: the jobs endpoint filters server-side, while the SQL one
     * returns every execution since start-up with its whole query plan attached, which is megabytes
     * to answer a question about the present.
     */
    public List<SparkJob> runningJobs() {
        Optional<String> app = applicationId();
        if (app.isEmpty()) {
            return List.of();
        }
        JsonNode listed = rest.get(url("/api/v1/applications/" + app.get() + "/jobs?status=running"));
        Instant now = clock.instant();
        List<SparkJob> jobs = new ArrayList<>();
        for (JsonNode job : listed) {
            jobs.add(new SparkJob(
                    job.path("jobId").asText(""),
                    text(job, "status", "RUNNING").toLowerCase(Locale.ROOT),
                    label(description(job)),
                    Stamps.ageS(text(job, "submissionTime", ""), now),
                    job.path("numCompletedTasks").asInt(0),
                    job.path("numTasks").asInt(0)));
        }
        return jobs;
    }

    /**
     * Kill the running jobs working on any of these statements, and report which.
     *
     * <p>Matched by statement rather than killing whatever is running, because the Thrift Server is
     * one application shared by everything that connects to it: a {@code spark-sql} session in the
     * container would otherwise be collateral damage when a comparison is cancelled.
     *
     * <p>Killing is necessary and not merely tidy. Taking a client's connection away stops this
     * backend waiting, but Spark carries on: HiveServer2 does not notice a dropped session promptly,
     * so the job keeps its share of the cores and the next comparison would be timed against an
     * orphan.
     */
    public List<String> killJobsFor(Collection<String> statements) {
        Set<String> wanted = new HashSet<>();
        for (String statement : statements) {
            wanted.add(label(statement));
        }
        List<String> killed = new ArrayList<>();
        for (SparkJob job : runningJobs()) {
            if (wanted.contains(job.sql())) {
                killJob(job.id());
                killed.add(job.id());
            }
        }
        return killed;
    }

    /**
     * Ask the UI to kill one job, as its own kill link does.
     *
     * <p>Depends on {@code spark.ui.killEnabled}, which is Spark's default and is left at it. The
     * handler answers with a redirect to the jobs page, so any 2xx or 3xx means the request was
     * accepted; whether the job dies is then Spark's business, and the next listing is what confirms
     * it.
     */
    public void killJob(String jobId) {
        HttpRequest.Builder request = rest.request(url("/jobs/job/kill/?id=" + jobId));
        HttpResponse<String> response =
                rest.send(request.POST(HttpRequest.BodyPublishers.noBody()).build());
        if (response.statusCode() >= 400) {
            throw new EngineFailed("the Spark UI refused to kill job " + jobId + " (HTTP "
                    + response.statusCode() + "); spark.ui.killEnabled must be on");
        }
    }

    /** A statement as the job list reports it, which is how one is matched against the other. */
    static String label(String statement) {
        return Messages.oneLine(statement, DESCRIPTION_LIMIT);
    }

    /**
     * {@code description} is the statement the Thrift Server set for the job group; {@code name} is
     * the Spark call site, which is all there is for a job submitted any other way.
     */
    private static String description(JsonNode job) {
        String described = text(job, "description", "");
        return described.isBlank() ? text(job, "name", "a Spark job") : described;
    }

    /**
     * A field's text, with an absent, null or empty field taking the fallback.
     *
     * <p>All three cases arrive: the UI omits {@code description} for a job nobody labelled, and
     * {@code JsonNode.asText} would answer the four characters "null" for a field explicitly null.
     */
    private static String text(JsonNode node, String field, String fallback) {
        JsonNode value = node.path(field);
        return value.isMissingNode() || value.isNull() || value.asText("").isEmpty()
                ? fallback
                : value.asText("");
    }

    private URI url(String path) {
        return URI.create(
                "http://%s:%d%s".formatted(settings.uiHost(), settings.appUiPort(), path));
    }
}
